docs/variants_from_haplotypes.tex

```diff
@@ -0,0 +1,66 @@
+\item Update insertion scores:  When we begin traversing row $i$ we initialize the current best score for alignments ending in a ``rightward'' gap (i.e. an insertion) as $R = -\infty$ and we initialize the length of the terminal gap in this best-scoring alignment as $S^r = 0$.  Note that these values are local to the inner loop over $j$ and thus do not need to be arrays\footnote{In the code, they \textit{are} arrays, but in the $i$th iteration of the loop over rows, only their $i$th elements are used.  Thus the elements are effectively scalars whose scope is the loop over $j$.}.  At each stage in the loop over $j$ the score for opening a rightward gap is $M_{i,j - 1} + w_{\rm open}$ and score for extending one is $R + w{\rm extend}$, where $R$ still pertains to the previous column $j - 1$.  We set $R$ (modifying it in-place)  to the greater of these values.  If the gap-opening score is greater than the gap-extending score, we set $S^r_j = 1$, otherwise we increment $S^r_j$ by 1.
+\item Record backtrack: The score for no gap, i.e. a match \textit{alignment} (as opposed to matching bases), is $M_{i-1,j-1}$ plus $w_{\rm match}$ if the $i$th reference base and $j$th alternate base agree or $w_{\rm mismatch}$ if they do not.  We now compare this score to the indel scores $D_j$ and $R$.  We set $B_{ij}$ to 0 if the match score is greatest, $S^d_j$ if the downward gap score is greatest, and $-S^r$ if the rightward gap score is greatest.  This convention essentially uses the sign an as enum in order to encode whether the optimal path has an insertion, deletion or match.
+\end{itemize}
+
+\section{Backward pass}
+After the forward pass the backtrack matrix $B$ is full.  We begin backtracking from the $(i,j)$ that is the maximum among the bottom row and rightmost column of $M$.  If this maximum is in the rightmost column that means all alternate haplotype bases are used and nothing special must be done.  If, however the maximum is on the bottom row it means that the more alternate bases remain.  In this case, we record a soft clip CIGAR element with length equal to the number of remaining bases at the end of the alignment\footnote{The code behaves differently for different values of the \code{OverhandStrategy} enum, but this strategy is hard-coded to \code{SOFTCLIP}, which results in the behavior we describe here.}.  Then, from this $(i,j)$ we iterate the following procedure until reaching $i = 0$ or $j = 0$ (recall that these correspond to immediately \textit{before} the start of the corresponding haplotypes): If $B_{ij} = 0$ add a match (M) CIGAR element to the left end of the alignment and move to $i,j \rightarrow i - 1, j - 1$.  If $B_{ij} = k > 0$ add a length-$k$ deletion CIGAR element to the left of the alignment and move to $i,j \rightarrow i - k, j$.  If $B_{ij} = -k < 0$ add a length-$k$ insertion CIGAR element to the left of the alignment and move to $i,j \rightarrow i - k, j$.  
```

Comment:
I liked that you added (M) to "add a match (M) CIGAR element..." Maybe you can do the same for deletion (D) CIGAR element, insertion (I) CIGAR element, and soft clip (S) CIGAR element?

Reply:
done